### PR TITLE
Switch caching strategy

### PR DESCRIPTION
**Pull request by phjudge**
This PR removes page caching from `bga_payroll/urls.py`, instead caching expensive queries per view.

We're shifting caching strategies in order to support a dynamic user login.

Remaining todos:
- (Not needed) ~Tackle template caching (e.g. in DepartmentView)~
- (Done) Refactor `views.py` to make cached properties less boilerplatey

### CODE REVIEW

**Comment by phjudge** [comment]
Per making the cached properties less boilerplatey...

What if we wrote a decorator to rewrite this...

```python
@property
def employee_salaries(self):
    data = self._cache.get('employee_salaries', [])

    if not data:
        data = self.object.employee_salaries
            
        cache.set('employee_salaries', data, CACHE_TIMEOUT)

    return data
```

...as this?

```python
@property
@use_cache(self._cache, CACHE_TIMEOUT)
def employee_salaries(self):
    return self.object.employee_salaries
```

File: payroll/views.py [hunk 1]
```diff
@@ -90,46 +115,88 @@ def get_context_data(self, **kwargs):
...
+        cache_prefix = str(self.object.id)
+        keys = [cache_prefix + '_' + k for k in cached_keys]
+
+        if not hasattr(self, '_kache'):
+            self._kache = cache.get_many(keys)
+
+        return self._kache
+
+    @property
+    @check_detail_cache
```
Comment (phjudge) [inline payroll/views.py]: This is the same as
```python
@check_cache(cache_prefix=lambda x: x.object.id)
```

File: payroll/views.py [hunk 2]
```diff
@@ -90,46 +115,88 @@ def get_context_data(self, **kwargs):
...
+            'expenditure_percentile': self.expenditure_percentile,
             'employee_salary_json': json.dumps(binned_employee_salaries),
             'data_year': 2017,
             'source_link': source_link,
         })
 
         return context
 
-    def median_entity_salary(self):
+    def get_median_entity_salary(self):
```
Comment (phjudge) [inline payroll/views.py]: Renamed the original function to a more explicit getter, used by the cached property.

Would it be cleaner to refactor this into the cached property itself?

File: payroll/views.py [hunk 3]
```diff
@@ -357,6 +424,31 @@ def highest_spending_department(self):
...
+        cached_keys = [
+            'aggregate_department_statistics',
+            'highest_spending_department',
+        ]
+        cache_prefix = str(self.object.id)
+        keys = [cache_prefix + '_' + k for k in cached_keys]
+
+        if not hasattr(self, '_kache'):
+            self._kache = super()._cache
+            self._kache.update(cache.get_many(keys))
```
Comment (phjudge) [inline payroll/views.py]: Because the `EmployerView` also implements `_cache`, we need to call both.

File: payroll/decorators.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    if func is None:
+        return partial(check_cache, cache_object=cache_object, cache_prefix=cache_prefix, cache_timeout=cache_timeout)
+
+    @wraps(func)
+    def _check_cache(self, *args, **kwargs):
+        # set key
+        if callable(cache_prefix):
+            key = cache_prefix(self) + '_' + func.__name__
+        elif cache_prefix:
+            key = cache_prefix + '_' + func.__name__
```
Comment (phjudge) [inline payroll/decorators.py]: We need to prefix our cache keys per view.

**Comment by phjudge** [comment]
@hancush This if functionally done now. There's a bit more potential refactoring to do.

I'd like your input on [pullrequestreview-277311050](#pullrequestreview-277311050), especially.

File: payroll/views.py (hunk 2 above)
Comment (hancush) [reply payroll/views.py]: i'm fine with it as written! it's nice to have the longer logic broken out.

**Comment by phjudge** [comment]
@hancush That's a great suggestion. I'll go ahead and implement that. Should we keep it in `views.py` or is it best practice to break it out into another file.

**Comment by hancush** [comment]
@phjudge there's a charts.py file with one charting mixin atm -- i wouldn't be opposed to converting that to a mixins.py file, if you don't mind updating the imports for the charting mixin. but it's up to you!

File: payroll/decorators.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        return partial(check_cache, cache_object=cache_object, cache_prefix=cache_prefix, cache_timeout=cache_timeout)
+
+    @wraps(func)
+    def _check_cache(self, *args, **kwargs):
+        # set key
+        if cache_prefix:
+            key = cache_prefix + '_' + func.__name__
+        elif getattr(self, 'cache_prefix'):
+            # get cache_prefix from CacheMixin attribute
+            key = getattr(self, 'cache_prefix') + '_' + func.__name__
```
Comment (phjudge) [inline payroll/decorators.py]: I take advantage of the `CacheMixin` attribute here as a fallback if no explicit `cache_prefix` is passed to the decorator.

File: payroll/mixins.py [hunk 6]
```diff
@@ -1,10 +1,42 @@
...
+
+    """
+
+    @property
+    def _cache(self):
+        if not hasattr(self, '_kache'):
+            if hasattr(self, 'cache_prefix'):
+                keys = [self.cache_prefix + '_' + k for k in self.cache_keys]
+            else:
+                keys = self.cache_keys
```
Comment (phjudge) [inline payroll/mixins.py]: When using class inheritance like we do in our views, there can only be one value for `cache_keys` — the child overrides the parent value. This means we can't use `super()` here. But we don't want it or need it!

File: payroll/views.py [hunk 7]
```diff
@@ -90,46 +117,84 @@ def get_context_data(self, **kwargs):
...
     template_name = 'unit.html'
 
+    def __init__(self):
+        super(EmployerView, self).__init__()
+        self.cache_keys += [
+            'aggregate_department_statistics',
+            'highest_spending_department',
+        ]
+        self.cache_keys = list(set(self.cache_keys))
+
```
Comment (phjudge) [inline payroll/views.py]: Extend the parent's `cache_keys`.

File: payroll/views.py [hunk 8]
```diff
@@ -90,46 +117,84 @@ def get_context_data(self, **kwargs):
...
         q = Salary.objects.filter(Q(job__position__employer__parent=self.object) | Q(job__position__employer=self.object))
 
         results = q.all().aggregate(median=Percentile('amount', 0.5, output_field=FloatField()))
 
         return results['median']
 
+    # CacheMixin attribute
+    @property
+    def cache_prefix(self):
+        return str(self.object.id)
```
Comment (phjudge) [inline payroll/views.py]: @hancush Is this sufficiently unique?, i.e. is this unique to all objects or just to objects of the same model?

**Comment by hancush** [comment]
We ultimately moved in a different direction. Closing.

**Pull request closed without merging**